File: go/day_8/main_test.go
package main

import (
	"reflect"
	"testing"
)

const (
	part1Solution = 284
	part2Solution = 973499
)

func TestPart1(t *testing.T) {
	tests := map[string]struct {
		data     []display
		expected int
	}{
		"example1": {
			data:     example1Data(),
			expected: 0,
		},
		"example2": {
			data:     example2Data(),
			expected: 26,
		},
		"actual": {
			data:     readData(),
			expected: part1Solution,
		},
	}

	for name, test := range tests {
		t.Run(name, func(t *testing.T) {
			actual := part1(test.data)

			if actual != test.expected {
				t.Fatalf("Expected: %v\nActual: %v", test.expected, actual)
			}
		})
	}
}

func TestPart2(t *testing.T) {
	tests := map[string]struct {
		data     []display
		expected int
	}{
		"example1": {
			data:     example1Data(),
			expected: 5353,
		},
		"example2": {
			data:     example2Data(),
			expected: 61229,
		},
		"actual": {
			data:     readData(),
			expected: part2Solution,
		},
	}

	for name, test := range tests {
		t.Run(name, func(t *testing.T) {
			actual := part2(test.data)

			if actual != test.expected {
				t.Fatalf("Expected: %v\nActual: %v", test.expected, actual)
			}
		})
	}
}

func TestParseContents(t *testing.T) {
	tests := map[string]struct {
		data     string
		expected []display
	}{
		"example1": {
			data:     "acedgfb cdfbe gcdfa fbcad dab cefabd cdfgeb eafb cagedb ab | cdfeb fcadb cdfeb cdbaf",
			expected: example1Data(),
		},
		"example2": {
			data: `be cfbegad cbdgef fgaecd cgeb fdcge agebfd fecdb fabcd edb | fdgacbe cefdb cefbgd gcbe
edbfga begcd cbg gc gcadebf fbgde acbgfd abcde gfcbed gfec | fcgedb cgb dgebacf gc
fgaebd cg bdaec gdafb agbcfd gdcbef bgcad gfac gcb cdgabef | cg cg fdcagb cbg
fbegcd cbd adcefb dageb afcb bc aefdc ecdab fgdeca fcdbega | efabcd cedba gadfec cb
aecbfdg fbg gf bafeg dbefa fcge gcbea fcaegb dgceab fcbdga | gecf egdcabf bgf bfgea
fgeab ca afcebg bdacfeg cfaedg gcfdb baec bfadeg bafgc acf | gebdcfa ecba ca fadegcb
dbcfg fgd bdegcaf fgec aegbdf ecdfab fbedc dacgb gdcebf gf | cefg dcbef fcge gbcadfe
bdfegc cbegaf gecbf dfcage bdacg ed bedf ced adcbefg gebcd | ed bcgafe cdgba cbgef
egadfb cdbfeg cegd fecab cgb gbdefca cg fgcdab egfdb bfceg | gbdfcae bgc cg cgb
gcafb gcf dcaebfg ecagb gf abcdeg gaef cafbge fdbac fegbdc | fgae cfgab fg bagce`,
			expected: example2Data(),
		},
	}

	for name, test := range tests {
		t.Run(name, func(t *testing.T) {
			actual := parseContents(test.data)

			if !reflect.DeepEqual(actual, test.expected) {
				t.Fatalf("Expected: %v\nActual: %v", test.expected, actual)
			}
		})
	}
}

func example1Data() []display {
	return []display{
		{
			digits: [10]uint8{
				0b1111111,
				0b0111110,
				0b1101101,
				0b0101111,
				0b0001011,
				0b0111111,
				0b1111110,
				0b0110011,
				0b1011111,
				0b0000011,
			},
			value: [4]uint8{
				0b0111110,
				0b0101111,
				0b0111110,
				0b0101111,
			},
		},
	}
}

func example2Data() []display {
	return []display{
		{
			digits: [10]uint8{
				0b0010010,
				0b1111111,
				0b1111110,
				0b1111101,
				0b1010110,
				0b1111100,
				0b1111011,
				0b0111110,
				0b0101111,
				0b0011010,
			},
			value: [4]uint8{
				0b1111111,
				0b0111110,
				0b1111110,
				0b1010110,
			},
		},
		{
			digits: [10]uint8{
				0b1111011,
				0b1011110,
				0b1000110,
				0b1000100,
				0b1111111,
				0b1111010,
				0b1101111,
				0b0011111,
				0b1111110,
				0b1110100,
			},
			value: [4]uint8{
				0b1111110,
				0b1000110,
				0b1111111,
				0b1000100,
			},
		},
		{
			digits: [10]uint8{
				0b1111011,
				0b1000100,
				0b0011111,
				0b1101011,
				0b1101111,
				0b1111110,
				0b1001111,
				0b1100101,
				0b1000110,
				0b1111111,
			},
			value: [4]uint8{
				0b1000100,
				0b1000100,
				0b1101111,
				0b1000110,
			},
		},
		{
			digits: [10]uint8{
				0b1111110,
				0b0001110,
				0b0111111,
				0b1011011,
				0b0100111,
				0b0000110,
				0b0111101,
				0b0011111,
				0b1111101,
				0b1111111,
			},
			value: [4]uint8{
				0b0111111,
				0b0011111,
				0b1111101,
				0b0000110,
			},
		},
		{
			digits: [10]uint8{
				0b1111111,
				0b1100010,
				0b1100000,
				0b1110011,
				0b0111011,
				0b1110100,
				0b1010111,
				0b1110111,
				0b1011111,
				0b1101111,
			},
			value: [4]uint8{
				0b1110100,
				0b1111111,
				0b1100010,
				0b1110011,
			},
		},
		{
			digits: [10]uint8{
				0b1110011,
				0b0000101,
				0b1110111,
				0b1111111,
				0b1111101,
				0b1101110,
				0b0010111,
				0b1111011,
				0b1100111,
				0b0100101,
			},
			value: [4]uint8{
				0b1111111,
				0b0010111,
				0b0000101,
				0b1111111,
			},
		},
		{
			digits: [10]uint8{
				0b1101110,
				0b1101000,
				0b1111111,
				0b1110100,
				0b1111011,
				0b0111111,
				0b0111110,
				0b1001111,
				0b1111110,
				0b1100000,
			},
			value: [4]uint8{
				0b1110100,
				0b0111110,
				0b1110100,
				0b1111111,
			},
		},
		{
			digits: [10]uint8{
				0b1111110,
				0b1110111,
				0b1110110,
				0b1111101,
				0b1001111,
				0b0011000,
				0b0111010,
				0b0011100,
				0b1111111,
				0b1011110,
			},
			value: [4]uint8{
				0b0011000,
				0b1110111,
				0b1001111,
				0b1110110,
			},
		},
		{
			digits: [10]uint8{
				0b1111011,
				0b1111110,
				0b1011100,
				0b0110111,
				0b1000110,
				0b1111111,
				0b1000100,
				0b1101111,
				0b1111010,
				0b1110110,
			},
			value: [4]uint8{
				0b1111111,
				0b1000110,
				0b1000100,
				0b1000110,
			},
		},
		{
			digits: [10]uint8{
				0b1100111,
				0b1100100,
				0b1111111,
				0b1010111,
				0b1100000,
				0b1011111,
				0b1110001,
				0b1110111,
				0b0101111,
				0b1111110,
			},
			value: [4]uint8{
				0b1110001,
				0b1100111,
				0b1100000,
				0b1010111,
			},
		},
	}
}

func BenchmarkReadData(b *testing.B) {
	for i := 0; i < b.N; i++ {
		if readData() == nil {
			b.FailNow()
		}
	}
}

func BenchmarkPart1(b *testing.B) {
	data := readData()
	for i := 0; i < b.N; i++ {
		if part1(data) != part1Solution {
			b.FailNow()
		}
	}
}

func BenchmarkPart2(b *testing.B) {
	data := readData()
	for i := 0; i < b.N; i++ {
		if part2(data) != part2Solution {
			b.FailNow()
		}
	}
}

File: ts/day_8/main.test.ts
import {
  assertEquals,
  assertNotEquals,
} from "https://deno.land/std@0.116.0/testing/asserts.ts";
import {
  bench,
  BenchmarkTimer,
  runBenchmarks,
} from "https://deno.land/std@0.116.0/testing/bench.ts";
import * as day from "./main.ts";

const part1Solution = 284;
const part2Solution = 973499;

Deno.test("part 1 real", () => {
  const input = day.readData();

  assertEquals(day.part1(input), part1Solution);
});

Deno.test("part 2 real", () => {
  const input = day.readData();

  assertEquals(day.part2(input), part2Solution);
});

Deno.test("parse contents example 1", () => {
  const input =
    `acedgfb cdfbe gcdfa fbcad dab cefabd cdfgeb eafb cagedb ab | cdfeb fcadb cdfeb cdbaf`;

  assertEquals(day.parseContents(input), example1Data());
});

Deno.test("parse contents example 2", () => {
  const input =
    `be cfbegad cbdgef fgaecd cgeb fdcge agebfd fecdb fabcd edb | fdgacbe cefdb cefbgd gcbe
edbfga begcd cbg gc gcadebf fbgde acbgfd abcde gfcbed gfec | fcgedb cgb dgebacf gc
fgaebd cg bdaec gdafb agbcfd gdcbef bgcad gfac gcb cdgabef | cg cg fdcagb cbg
fbegcd cbd adcefb dageb afcb bc aefdc ecdab fgdeca fcdbega | efabcd cedba gadfec cb
aecbfdg fbg gf bafeg dbefa fcge gcbea fcaegb dgceab fcbdga | gecf egdcabf bgf bfgea
fgeab ca afcebg bdacfeg cfaedg gcfdb baec bfadeg bafgc acf | gebdcfa ecba ca fadegcb
dbcfg fgd bdegcaf fgec aegbdf ecdfab fbedc dacgb gdcebf gf | cefg dcbef fcge gbcadfe
bdfegc cbegaf gecbf dfcage bdacg ed bedf ced adcbefg gebcd | ed bcgafe cdgba cbgef
egadfb cdbfeg cegd fecab cgb gbdefca cg fgcdab egfdb bfceg | gbdfcae bgc cg cgb
gcafb gcf dcaebfg ecagb gf abcdeg gaef cafbge fdbac fegbdc | fgae cfgab fg bagce`;

  assertEquals(day.parseContents(input), example2Data());
});

Deno.test("part 1 example 1", () => {
  const input = example1Data();

  assertEquals(day.part1(input), 0);
});

Deno.test("part 1 example 2", () => {
  const input = example2Data();

  assertEquals(day.part1(input), 26);
});

Deno.test("part 2 example 1", () => {
  const input = example1Data();

  assertEquals(day.part2(input), 5353);
});

Deno.test("part 2 example 2", () => {
  const input = example2Data();

  assertEquals(day.part2(input), 61229);
});

const example1Data = () => [
  new day.Display().fromComponents([
    "acedgfb",
    "cdfbe",
    "gcdfa",
    "fbcad",
    "dab",
    "cefabd",
    "cdfgeb",
    "eafb",
    "cagedb",
    "ab",
  ], ["cdfeb", "fcadb", "cdfeb", "cdbaf"]),
];

const example2Data = () => [
  new day.Display().fromComponents([
    "be",
    "cfbegad",
    "cbdgef",
    "fgaecd",
    "cgeb",
    "fdcge",
    "agebfd",
    "fecdb",
    "fabcd",
    "edb",
  ], ["fdgacbe", "cefdb", "cefbgd", "gcbe"]),
  new day.Display().fromComponents([
    "edbfga",
    "begcd",
    "cbg",
    "gc",
    "gcadebf",
    "fbgde",
    "acbgfd",
    "abcde",
    "gfcbed",
    "gfec",
  ], ["fcgedb", "cgb", "dgebacf", "gc"]),
  new day.Display().fromComponents([
    "fgaebd",
    "cg",
    "bdaec",
    "gdafb",
    "agbcfd",
    "gdcbef",
    "bgcad",
    "gfac",
    "gcb",
    "cdgabef",
  ], ["cg", "cg", "fdcagb", "cbg"]),
  new day.Display().fromComponents([
    "fbegcd",
    "cbd",
    "adcefb",
    "dageb",
    "afcb",
    "bc",
    "aefdc",
    "ecdab",
    "fgdeca",
    "fcdbega",
  ], ["efabcd", "cedba", "gadfec", "cb"]),
  new day.Display().fromComponents([
    "aecbfdg",
    "fbg",
    "gf",
    "bafeg",
    "dbefa",
    "fcge",
    "gcbea",
    "fcaegb",
    "dgceab",
    "fcbdga",
  ], ["gecf", "egdcabf", "bgf", "bfgea"]),
  new day.Display().fromComponents([
    "fgeab",
    "ca",
    "afcebg",
    "bdacfeg",
    "cfaedg",
    "gcfdb",
    "baec",
    "bfadeg",
    "bafgc",
    "acf",
  ], ["gebdcfa", "ecba", "ca", "fadegcb"]),
  new day.Display().fromComponents([
    "dbcfg",
    "fgd",
    "bdegcaf",
    "fgec",
    "aegbdf",
    "ecdfab",
    "fbedc",
    "dacgb",
    "gdcebf",
    "gf",
  ], ["cefg", "dcbef", "fcge", "gbcadfe"]),
  new day.Display().fromComponents([
    "bdfegc",
    "cbegaf",
    "gecbf",
    "dfcage",
    "bdacg",
    "ed",
    "bedf",
    "ced",
    "adcbefg",
    "gebcd",
  ], ["ed", "bcgafe", "cdgba", "cbgef"]),
  new day.Display().fromComponents([
    "egadfb",
    "cdbfeg",
    "cegd",
    "fecab",
    "cgb",
    "gbdefca",
    "cg",
    "fgcdab",
    "egfdb",
    "bfceg",
  ], ["gbdfcae", "bgc", "cg", "cgb"]),
  new day.Display().fromComponents([
    "gcafb",
    "gcf",
    "dcaebfg",
    "ecagb",
    "gf",
    "abcdeg",
    "gaef",
    "cafbge",
    "fdbac",
    "fegbdc",
  ], ["fgae", "cfgab", "fg", "bagce"]),
];

bench({
  name: "read data",
  runs: 500,
  func(b: BenchmarkTimer): void {
    b.start();
    const input = day.readData();
    assertNotEquals(input, []);
    b.stop();
  },
});

bench({
  name: "part 1",
  runs: 500,
  func(b: BenchmarkTimer): void {
    const input = day.readData();
    b.start();
    assertEquals(day.part1(input), part1Solution);
    day.part1(input);
    b.stop();
  },
});

bench({
  name: "part 2",
  runs: 500,
  func(b: BenchmarkTimer): void {
    const input = day.readData();
    b.start();
    assertEquals(day.part2(input), part2Solution);
    b.stop();
  },
});

if (Deno.args.length > 0 && Deno.args[0] == "--bench") {
  runBenchmarks();
}
